modules/gdscript/README.md

```diff
@@ -0,0 +1,92 @@
+# Basic GDScript module architecture
+This provides some basic information in how GDScript is implemented and integrates with the rest of the engine. You can learn more about GDScript in the [documentation](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html). It describes the syntax and user facing systems and concepts, and can be used as a reference for what user expectations are.
+
+
+## General design
+
+GDScript is:
+
+1. A dynamically typed language. Type hints help with static analysis and performance, but typed code must easily interoperate with untyped code.
+2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.
+3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, [C#](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
+
+
+## Compilation
+Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.
+
+The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a class defined in GDScript. Each `.gd` file is called a _class file_ because it implicitly defines a class in GDScript, and thus results in an associated `GDScript` object. However, GDScript classes may define [_inner classes_](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inner-classes), and those are also represented by further `GDScript` objects, even though they are not in files of their own.
+
+The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.
+
+A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.
+
+### Loading source code
+
+This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!
+
+### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))
+
+Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.
```

Comment:
I've actually very rarely touched the tokenizer and parser, so I trust your expertise here. What I noticed is that references to the tokenizer only ever shows up in the 1) parser, 2) gdscript editor, 3) highlighter, and 4) test framework. I think you're right though, the highlighter doesn't really seem to use it a lot, just references some tokens.

Reply:
I rewrote this a little bit, because the exception was meant to say that we don't see it independent of the parser, except in e.g. the code editor (which is one of the rare exceptions) :)